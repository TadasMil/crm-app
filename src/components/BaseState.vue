<script setup name="BaseState" lang="ts">
import { injectState } from '../composables/useState'

import BaseLoading from './BaseLoading.vue'
import BaseSuccess from './BaseSuccess.vue'

interface Props {
  injectKey?: string
  successTitle?: string
}

const props = withDefaults(defineProps<Props>(), {
  injectKey: 'state',
  successTitle: '',
})

const { stateIsLoading, stateIsIdle, stateIsSuccess } = injectState(
  props.injectKey
)
</script>

<template>
  <div v-if="stateIsIdle">
    <slot />
  </div>

  <slot v-else-if="stateIsLoading" name="loading">
    <BaseLoading />
  </slot>

  <slot v-else-if="stateIsSuccess" name="success">
    <BaseSuccess :title="successTitle">
      <slot name="success-actions" />
    </BaseSuccess>
  </slot>
</template>
