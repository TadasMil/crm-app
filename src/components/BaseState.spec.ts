import { render, screen } from '@testing-library/vue'
import { describe, expect, it } from 'vitest'

import BaseState from './BaseState.vue'

describe('BaseState.spec.js', () => {
  it('should render the component with idle state', () => {
    const { html } = render(BaseState, {
      global: {
        provide: {
          state: {
            value: 'IDLE',
          },
        },
      },
      slots: {
        default: '<div>im idle slot</div>',
      },
    })

    expect(html()).toMatchSnapshot()
  })

  it('should render the component with loading state', () => {
    const { html } = render(BaseState, {
      global: {
        stubs: ['BaseLoading'],
        provide: {
          state: {
            value: 'LOADING',
          },
        },
      },
    })

    expect(html()).toMatchSnapshot()
  })

  it('should render the component with loading state and loading slot', () => {
    const { html } = render(BaseState, {
      global: {
        stubs: ['BaseLoading'],
        provide: {
          state: {
            value: 'LOADING',
          },
        },
      },
      slots: {
        loading: '<div>im loading slot</div>',
      },
    })

    expect(html()).toMatchSnapshot()
  })

  it('should render the component with success state', () => {
    const { html } = render(BaseState, {
      global: {
        stubs: ['BaseSuccess'],
        provide: {
          state: {
            value: 'SUCCESS',
          },
        },
      },
    })

    expect(html()).toMatchSnapshot()
  })

  it('should render the component with success state', () => {
    const { html } = render(BaseState, {
      global: {
        stubs: ['BaseSuccess'],
        provide: {
          state: {
            value: 'SUCCESS',
          },
        },
      },
    })

    expect(html()).toMatchSnapshot()
  })

  it('should render the component with success state and success slot', () => {
    const { html } = render(BaseState, {
      global: {
        stubs: ['BaseSuccess'],
        provide: {
          state: {
            value: 'SUCCESS',
          },
        },
      },
      slots: {
        success: '<div>success slot</div>',
      },
    })

    expect(html()).toMatchSnapshot()
  })
})
