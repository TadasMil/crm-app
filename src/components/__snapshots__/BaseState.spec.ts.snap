// Vitest Snapshot v1

exports[`BaseState.spec.js > should render the component with idle state 1`] = `
"<div>
  <div>im idle slot</div>
</div>"
`;

exports[`BaseState.spec.js > should render the component with loading state 1`] = `"<base-loading-stub title=\\"We're loading data.. Please wait\\"></base-loading-stub>"`;

exports[`BaseState.spec.js > should render the component with loading state and loading slot 1`] = `"<div>im loading slot</div>"`;

exports[`BaseState.spec.js > should render the component with success state 1`] = `"<base-success-stub title=\\"\\"></base-success-stub>"`;

exports[`BaseState.spec.js > should render the component with success state 2`] = `"<base-success-stub title=\\"\\"></base-success-stub>"`;

exports[`BaseState.spec.js > should render the component with success state and success slot 1`] = `"<div>success slot</div>"`;
